app/controllers/application_controller.rb

```diff
@@ -2,4 +2,18 @@
 
 class ApplicationController < ActionController::Base
   include Localization
+
+  helper_method :current_user
+
+  private
+
+  def current_user
+    return unless session[:user_id]
+
+    User.find(session[:user_id])
```

Comment:
Do we need to use instance variable in this case?

Reply:
sure, instance variable give the class instance (object) its own private space to store data as you already know - so its value is accesible by every methods of the object. 
besides that, the purpose of momoization is to cache the expensive calculation so we need to do it once only. If we dont use instance variable to store the computation result - in our case, it's the DB query result - but using a normal variable only, once the method completes it job, the value is gone, no caching envolved. 
here we are storing the value in the instance variable, so as long as the object still exists, any other methods call `current_user`, it will check the value of the instance varible, since the value already exist & thanks to the operator `=||`, no DB query is necessary anymore.
I hope this article give you more insight https://www.honeybadger.io/blog/ruby-rails-memoization/

but the code I suggested above isn't good in case the `user_id` doesn't exist. In that case, @current_user is always `nil` so the DB query is repeated every time the method is called. It should be
```ruby
def current_user
  return @current_user if @current_user_checked

  @current_user_checked = true
  @current_user =|| User.find_by_id(session[:user_id])
end
```
or better:
```ruby
def logged_in?
   sesson[:user_id].present?
end

def current_user
  return unless logged_in?
  @current_user =|| User.find_by_id(session[:user_id])
end
```
by this way, in the before_action you're using, you can change the checking so it's clearer, instead of
```ruby
before_action :redirect_to_home, if: :current_user, only: %i[new create]
```
it becomes
```
before_action :redirect_to_home, if: :logged_in?, only: %i[new create]
```

Reply:
@longnd Thank you so much for an explanation. Will make the change.

Reply:
Fixed in f3b23e34cabed85b3efbb11273a20ca509c0ed36